plugins/notifications-backend-module-slack/src/lib/SlackNotificationProcessor.ts

```diff
@@ -0,0 +1,285 @@
+          );
+          failureCount++;
+        }
+      });
+
+      this.messagesSent.add(successCount);
+      this.messagesFailed.add(failureCount);
+    };
+  }
+
+  getName(): string {
+    return 'SlackNotificationProcessor';
+  }
+
+  async processOptions(
+    options: NotificationSendOptions,
+  ): Promise<NotificationSendOptions> {
+    if (options.recipients.type !== 'entity') {
+      return options;
+    }
+
+    const entityRefs = [options.recipients.entityRef].flat();
+
+    const outbound: ChatPostMessageArguments[] = [];
+    await Promise.all(
+      entityRefs.map(async entityRef => {
+        const compoundEntityRef = parseEntityRef(entityRef);
+        // skip users as they are sent direct messages, but allow all other entity kinds
+        // to have a channel id annotation.
+        if (compoundEntityRef.kind === 'user') {
+          return;
+        }
+
+        let channel;
+        try {
+          channel = await this.getChannelId(entityRef);
+        } catch (error) {
+          this.logger.error(
+            `Failed to get Slack channel for entity: ${
+              (error as Error).message
+            }`,
+          );
+          return;
+        }
+
+        if (!channel) {
+          this.logger.debug(`No Slack channel found for entity: ${entityRef}`);
+          return;
+        }
+
+        this.logger.debug(
+          `Sending notification with payload: ${JSON.stringify(
+            options.payload,
+          )}`,
+        );
+
+        const payload = toChatPostMessageArgs({
+          channel,
+          payload: options.payload,
+        });
+
+        this.logger.debug(
+          `Sending Slack channel notification: ${JSON.stringify(payload)}`,
+        );
+        outbound.push(payload);
+      }),
+    );
+
+    console.log('dispatching message');
+    await this.sendNotifications(outbound);
+
+    return options;
+  }
+
+  async postProcess(
+    notification: Notification,
+    options: NotificationSendOptions,
+  ): Promise<void> {
+    if (options.recipients.type === 'broadcast' || !notification.user) {
+      return;
+    }
+
+    const entityRefs = [options.recipients.entityRef].flat();
+    if (entityRefs.some(e => parseEntityRef(e).kind === 'group')) {
+      // We've already dispatched a slack channel message, so let's not send a DM.
+      return;
+    }
+
+    const destination = await this.getSlackUserId(notification.user);
+    if (!destination) {
+      this.logger.error(`No email found for user entity: ${notification.user}`);
+      return;
+    }
+
+    const payload = toChatPostMessageArgs({
+      channel: destination,
+      payload: options.payload,
+    });
+
+    this.logger.debug(`Sending DM notification: ${JSON.stringify(payload)}`);
+
+    // batch it up
+    await this.sendNotifications([payload]);
+  }
+
+  async getEntities(
+    entityRefs: readonly string[],
+  ): Promise<(Entity | undefined)[]> {
+    const { token } = await this.auth.getPluginRequestToken({
```

Comment:
Could probably use cacheService here (and other entity getters) like in email processor to prevent unnecessary calls to catalog. But can be added later if this becomes an issue or WDYT?